Localization/cubature_kalman_filter/cubature_kalman_filter.py

```diff
@@ -0,0 +1,211 @@
+        if i == (N - 1) and show_final == 1:
+            show_final_flag = 1
+        else:
+            show_final_flag = 0
+        if show_animation == 1:
+            plot_animation(i, x_true_cat, x_est_cat, z)
+        if show_ellipse == 1:
+            plot_ellipse(x_est[0:2], p_est)
+        if show_final_flag == 1:
+            plot_final(x_true_cat, x_est_cat, z_cat)
+        x_est, p_est = cubature_kalman_filter(x_est, p_est, z)
+        x_true_cat = np.vstack((x_true_cat, x_true[0:2].T))
+        x_est_cat = np.vstack((x_est_cat, (x_est[0:2].T)))
+        z_cat = np.vstack((z_cat, z[0:2].T))
+    print('CKF Over')
+
+
+def cubature_kalman_filter(x_est, p_est, z):
+    x_pred, p_pred = cubature_prediction(x_est, p_est)
+    x_upd, p_upd = cubature_update(x_pred, p_pred, z)
+    return x_upd.astype(float), p_upd.astype(float)
+
+
+def f(x):
+    x[0] = x[0] + (x[3]/x[4]) * (np.sin(x[4] * dt + x[2]) - np.sin(x[2]))
+    x[1] = x[1] + (x[3]/x[4]) * (- np.cos(x[4] * dt + x[2]) + np.cos(x[2]))
+    x[2] = x[2] + x[4] * dt
+    x[3] = x[3]
+    x[4] = x[4]
+    x.reshape((5, 1))
+    return x.astype(float)
+
+
+def h(x):
+    x = hx @ x
+    x.reshape((4, 1))
+    return x
+
+
+def sigma(x, p):
+    n = np.shape(x)[0]
+    SP = np.zeros((n, 2*n))
+    W = np.zeros((1, 2*n))
+    for i in range(n):
+        SD = sqrtm(p)
+        SP[:, i] = (x + (math.sqrt(n) * SD[:, i]).reshape((n, 1))).flatten()
+        SP[:, i+n] = (x - (math.sqrt(n) * SD[:, i]).reshape((n, 1))).flatten()
+        W[:, i] = 1/(2*n)
+        W[:, i+n] = W[:, i]
+    return SP.astype(float), W.astype(float)
+
+
+def cubature_prediction(x_pred, p_pred):
+    n = np.shape(x_pred)[0]
+    [SP, W] = sigma(x_pred, p_pred)
+    x_pred = np.zeros((n, 1))
+    p_pred = q
+    for i in range(2*n):
+        x_pred = x_pred + (f(SP[:, i]).reshape((n, 1)) * W[0, i])
+    for i in range(2*n):
+        p_step = (f(SP[:, i]).reshape((n, 1)) - x_pred)
+        p_pred = p_pred + (p_step @ p_step.T * W[0, i])
+    return x_pred.astype(float), p_pred.astype(float)
+
+
+def cubature_update(x_pred, p_pred, z):
+    n = np.shape(x_pred)[0]
+    m = np.shape(z)[0]
+    [SP, W] = sigma(x_pred, p_pred)
+    y_k = np.zeros((m, 1))
+    P_xy = np.zeros((n, m))
+    s = r
+    for i in range(2*n):
+        y_k = y_k + (h(SP[:, i]).reshape((m, 1)) * W[0, i])
+    for i in range(2*n):
+        p_step = (h(SP[:, i]).reshape((m, 1)) - y_k)
+        P_xy = P_xy + ((SP[:, i]).reshape((n, 1)) -
+                       x_pred) @ p_step.T * W[0, i]
+        s = s + p_step @ p_step.T * W[0, i]
+    x_pred = x_pred + P_xy @ np.linalg.pinv(s) @ (z - y_k)
+    p_pred = p_pred - P_xy @ np.linalg.pinv(s) @ P_xy.T
+    return x_pred, p_pred
+
+
+def linear_update(x_pred, p_pred, z):
+    s = h @ p_pred @ h.T + r
+    k = p_pred @ h.T @ np.linalg.pinv(s)
+    v = z - h @ x_pred
+
+    x_upd = x_pred + k @ v
+    p_upd = p_pred - k @ s @ k.T
+    return x_upd.astype(float), p_upd.astype(float)
```

Comment:
I have not used this function in the code, I think I forgot to remove it after debugging. So I have removed it in the latest commit.